DM-15675: improve data ID expansion and filename templates

TallJimbo opened the pull request.

TallJimbo committed:
Use valid dimensions in templates and tests.

Most of these tests don't actually depend on having the right
dimensions, but they may share configuration with tests that do.
Using valid dimensions also mitigates damage from cargo-culting
and confusion from reading the tests.

TallJimbo committed:
Test that template expansion is minimal over dimensions.

Either minimal or maximal would work; what matters is that the template
doesn't depend on whether the data ID has extra keys or not.

TallJimbo committed:
Improve data ID metadata expansion.

This adds an option to expand the data ID entries to include links
for all optional dependencies, as well as an option (now default) to
only query when the requested values are missing.

TallJimbo committed:
Expand to include optional-dependency links in addDataset.

TallJimbo committed:
Change nomenclature from 'optional' to 'implied' dependencies.

TallJimbo committed:
Convert DimensionGraph properties into methods.

The 'implied' property is now a method with an optional argument, and
the old 'asSet` property has been renamed to 'toSet' and is now a
method, to emphasize that the fact that it returns an internal object
is an implementation detail, not part of the interface.

TallJimbo committed:
Standardize APIs for accessing dimension links.

DimensionElement, DimensionSet, and DimensionGraph now all have a
'links' method instead of the mishmash of differently-named properties
they had before.

This also mostly removes the term "primary key" from the documentation
about links; I've left it in a couple of places because it still seems
like the best way to describe the 'link' concept to a complete
newcomer to the codebase.

TallJimbo committed:
Further rewrite of data ID expansion.

This adds support for expanding data IDs to include links for implied
dependencies recursively.  For example, given a data ID with a "visit"
key, we can now lookup the "physical_filter" key from the "Visit" table
and the "abstract_filter" key from the "AbstractFilter" table.

TallJimbo committed:
Doc fixes.

natelust commented on python/lsst/daf/butler/registries/sqlRegistry.py:
```diff
@@ -347,6 +347,22 @@ def addDataset(self, datasetType, dataId, run, producer=None, recursive=False, *
+                try:
+                    links[link] = dataId.entries[dim][link]
+                except KeyError:
+                    raise LookupError(f"'{link}' not provided in data ID, but Registry is limited.")
```
This could be a more helpful error message

TallJimbo replied:
This particular error message was removed in a later commit on this branch, but I've improved the error message that occurs for the same condition there.

natelust commented on python/lsst/daf/butler/core/dimensions/dataId.py:
```diff
@@ -205,7 +206,7 @@ def __init__(self, dataId=None, *, dimensions=None, dimension=None, universe=Non
                 self.entries[element].update(subdict)
 
-        missing = self.dimensions.links - self._links.keys()
+        missing = self.dimensions.links() - self._links.keys()
```
I think it is a slight shame that links is a method on dimensions in one case and _links is a dict on self, but I am not sure I have something much better to suggests, except perhaps longer names?

TallJimbo replied:
`self.dimensions.links()` is also a `frozenset` while `self._links` is a `dict`.  I'll rename the latter to `_linkValues`.

TallJimbo committed:
Improve error message in disableWhenLimited decorator.

Since this may now be a much more common failure when expanding data
IDs, it makes sense to expand the error message for that case.

TallJimbo committed:
Rename DataId._links to _linkValue.

This makes a bit more clear the distinction between this dict
and the links() methods on various other dimension objects.
